### ISSUE TITLE
Core service not starting after restart

### ISSUE THREAD

**Issue by ghost** [safing/portmaster#482]
**Pre-Submit Checklist**:

- Check applicable sources for existing issues:
  - [Windows Known Issues](https://docs.safing.io/portmaster/install/windows#known-issues)
  - [Linux Known Issues](https://docs.safing.io/portmaster/install/linux#compatibility)
  - [Github Issues](https://github.com/safing/portmaster/issues?q=is%3Aissue+label%3Abug)

**What happened**:
After installing the portmaster, it would work as expected, but after a computer restart it would ask for a password to start the core service but would never start after the password was given

**What did you expect to happen?**:
for the core service to start after the password was inputted

**How did you reproduce it?**:
1) uninstall the portmaster
2) install the portmaster rpm with zypper
3) run the portmaster and input the password
4) restart the computer
5) run the portmaster and enter the password
6) see the issue

**Debug Information**:

Version: 0.7.12
OS: openSUSE Tumbleweed 20211220
Kernel: 5.15.8-1

**Comment by dhaavi** [comment]
Hey @BobIsMyManager, thanks for reporting this!

It sounds like the core service is not starting on boot and when triggered by the UI. Can you paste the output of `systemctl status portmaster`?

**Comment by BobIsMyManager** [comment]
`Unit portmaster.service could not be found.` even though the app does start on boot, just not the core service

**Comment by dhaavi** [comment]
It seems it hasn't installed correctly.
Have you tried removing and installing again?

**Comment by BobIsMyManager** [comment]
yes and it works when i reinstall it until my pc restarts where it stops working again

**Comment by BobIsMyManager** [comment]
after i install it it says 
`Additional rpm output:
portmaster: Configuring portmaster.service to launch at boot
Created symlink /etc/systemd/system/multi-user.target.wants/portmaster.service -> /opt/safing/portmaster/portmaster.service.
Created symlink /etc/systemd/system/portmaster.service -> /opt/safing/portmaster/portmaster.service.
chcon: can't apply partial context to unlabeled file '/opt/safing/portmaster/portmaster-start'
warning: %post(portmaster-0:0.7.0~1-1.x86_64) scriptlet failed, exit status 1`

**Comment by VottonDev** [comment]
Sounds like chcon is applying the wrong SELinux context, so probably can try manually changing it to see if it works or possibly using semanage to relabel the file.

Haven't particularly dealt with openSUSE, but could possibly help you.

**Comment by dhaavi** [comment]
We are currently investigating some SELinux stuff, hopefully we'll have an update on this soon.

**Comment by BobIsMyManager** [comment]
I also now seem to have an issue where similarly to #487, the GUI of the app wont start. The output of trying to start the app with `/opt/safing/portmaster/portmaster-start app` is: `Error: please set the data directory using --data=/path/to/data/dir`

**Comment by BobIsMyManager** [comment]
And once i set the data directory using `sudo /opt/safing/portmaster/portmaster-start --data=/opt/safing/portmaster app`, this is what happens

```
[control] 2022/02/06 21:11:39 app/portmaster-app.zip failed with: could not get component: the requested file is not available locally
[control] 2022/02/06 21:11:41 app/portmaster-app.zip failed with: could not get component: the requested file is not available locally
[control] 2022/02/06 21:11:45 updating registry index
[control] 2022/02/06 21:11:45 app/portmaster-app.zip failed with: could not get component: the requested file is not available locally
[control] 2022/02/06 21:11:51 updating registry index
[control] 2022/02/06 21:11:51 app/portmaster-app.zip failed with: could not get component: the requested file is not available locally
[control] 2022/02/06 21:11:59 updating registry index
[control] 2022/02/06 21:11:59 app/portmaster-app.zip failed with: could not get component: the requested file is not available locally
[control] 2022/02/06 21:11:59 encountered 5 consecutive errors, giving up ...
Error: could not get component: the requested file is not available locally
```

**Comment by BobIsMyManager** [comment]
Ive managed to collect more info to try help you guys out:

1. After starting the portmaster, `Error: connect ECONNREFUSED 127.0.0.1:817` keeps appearing in the terminal infinitely

2. After pressing start core service and entering the root password, this error appears
```
Error occurred in handler for 'api.startService': {
  error: Error: Command failed: cd "/home/BIMM"; "/usr/bin/pkexec" --disable-internal-agent /bin/bash -c "echo SUDOPROMPT; systemctl start portmaster.service --no-pager"
  Failed to start portmaster.service: Unit portmaster.service not found.

      at ChildProcess.exithandler (child_process.js:317:12)
      at ChildProcess.emit (events.js:315:20)
      at maybeClose (internal/child_process.js:1048:16)
      at Socket.<anonymous> (internal/child_process.js:439:11)
      at Socket.emit (events.js:315:20)
      at Pipe.<anonymous> (net.js:673:12) {
    killed: false,
    code: 5,
    signal: null,
    cmd: 'cd "/home/BIMM"; "/usr/bin/pkexec" --disable-internal-agent /bin/bash -c "echo SUDOPROMPT; systemctl start portmaster.service --no-pager"'
  },
  stdout: '',
  stderr: 'Failed to start portmaster.service: Unit portmaster.service not found.\n'
}
```

3. When the portmaster starts and when the root password is entered, this appears
```
(portmaster-app_v0-2-2:2972): LIBDBUSMENU-GLIB-WARNING **: 22:13:22.002: About to Show called on an item wihtout submenus.  We're ignoring it.
```

Hopefully this can help you guys figure out whats causing this issue 😃

**Comment by dhaavi** [comment]
It looks like you're still missing the `portmaster.service` file.
Can you try following the manual install instructions from [here](https://docs.safing.io/portmaster/install/linux#manual-install-and-launching)?

**Comment by BobIsMyManager** [comment]
This is what happens when i run `sudo /opt/safing/portmaster/portmaster-start core`. Here is a snippet of what is being spammed in console

```
220211 17:41:04.685 les/worker:098 ▶ INFO 620 nameserver: service-worker dns resolver requested restart: stopped conflicting name service with pid 718 - restarting now
220211 17:41:04.685 r/takeover:118 ▶ WARN 621 nameserver: killed conflicting service with PID 718 over [::1]:53
```

https://user-images.githubusercontent.com/70704875/153641629-378e4275-4932-4419-a665-b8dc67dcaec6.mov

**Comment by dhaavi** [comment]
It looks like you have another service running on port 53 - is it possible to disable that?

Also, the Portmaster should not go into a loop like this, will investigate.

**Comment by BobIsMyManager** [comment]
I disabled the service on port 53 and now things seem to work now!
![Screenshot_20220211_175935](https://user-images.githubusercontent.com/70704875/153644583-8c665ba7-f831-4e30-8e41-0f2179457c79.png)

**Comment by BobIsMyManager** [comment]
It seems that sometimes when I start the core service by pressing the button on the portmaster, it doesnt work, but when I start it manually in the command line, it does work.

**Comment by BobIsMyManager** [comment]
It seems that when I start the core service by pressing the button on the portmaster, it doesnt work, but when I start it manually in the command line, it does work.

**Comment by BobIsMyManager** [comment]
Another weird bug that seems to be related to this issue:

![20220326_12h58m24s_grim](https://user-images.githubusercontent.com/70704875/160281620-6d37a8ef-39a0-4d2b-a688-186139eeeac9.png)

And when trying to execute `sudo chcon -t bin_t /opt/safing/portmaster/portmaster-start` to fix this according to the portmaster docs, I get this error:
```
chcon: can't apply partial context to unlabeled file '/opt/safing/portmaster/portmaster-start'
```

**Comment by dhaavi** [comment]
> I disabled the service on port 53 and now things seem to work now!

Great!

> It seems that when I start the core service by pressing the button on the portmaster, it doesnt work, but when I start it manually in the command line, it does work.

How are you trying to start it?

> Another weird bug that seems to be related to this issue (this happens every time I restart my system):

The Portmaster is a system service, so you might need to use `sudo systemctl status portmaster`.

> And when trying to execute sudo chcon -t bin_t /opt/safing/portmaster/portmaster-start to fix this according to the portmaster docs, I get this error:

We are still getting this (SELinux) sorted out and we will update the docs when ready.

**Comment by BobIsMyManager** [comment]
> How are you trying to start it?

When starting the core service with `sudo /opt/safing/portmaster/portmaster-start core` or `sudo systemctl enable --now portmaster` it works as intended, but when trying to start it by pressing the button below, nothing happens.
![20220328_17h50m11s_grim](https://user-images.githubusercontent.com/70704875/160448281-777a4f82-4c74-48f2-a9c4-5a34d10709ad.png)
 
> The Portmaster is a system service, so you might need to use sudo systemctl status portmaster.

The same thing still happens
![20220328_17h51m38s_grim](https://user-images.githubusercontent.com/70704875/160448427-70fdb4d3-2710-47ef-bd4d-b73392470406.png)

**Comment by ppacher** [comment]
Hi @BobIsMyManager,
if you press the "Start Core Service" button in the UI it tries to start the Portmaster using systemd but requires a polkit-agent to be running so you can be prompted for your "sudo" password. Unfortunately there's no easy way to detect if a polkit-agent is running (because there are so many of them and some DE's directly integrate that). 

Seems like you are running a window manager similar to i3, dwm or awesome. They normally don't provide a polkit-agent out of the box but you can just start it yourself. For example, if you install the polkit-gnome package you can start the agent using `/usr/lib/polkit-gnome/polkit-gnome-authentication-agent-1`.

I will talk to the team on how we could improve that situation. Maybe adding some help-text to the "Portmaster Core Service is not running" screen would be enough for now.

**Comment by BobIsMyManager** [comment]
> Can you post the output of ls -lah /etc/systemd/system/portmaster.service please?

`lrwxrwxrwx 1 root root 41 Mar 25 17:56 /etc/systemd/system/portmaster.service -> /opt/safing/portmaster/portmaster.service`

**Comment by BobIsMyManager** [comment]
Now when executing `sudo systemctl --user start portmaster` I get the response: `Failed to connect to bus: No medium found`

**Comment by ppacher** [comment]
Is this still an issue?

**Comment by somerandomguy24** [comment]
Yes, some things have changed over time so I'll get back to you on that.

**Comment by vaardan** [comment]
Re-opening the issue, as I've encountered it as well.

**Summary**
For some unknown reason unit `portmaster.service` doesn't exist when executing `systemctl start portmaster` or `systemctl status portmaster`.

But it does exist when executing `systemctl enable portmaster`. After executing `systemctl enable`, both `systemctl start` and `systemctl status` see the unit and work as expected. 

After the restart, autostart doesn't happen and `portmaster.service` is once again is "gone" until `systemctl enable` is executed.

**System and installation info**
OS: **Fedora Linux 40 (Workstation Edition) x86_64**
Kernel: **6.11.10-200.fc40.x86_64**
Shell: **zsh 5.9**
DE: **GNOME 46.6**
Display server: **Wayland**
Portmaster version: **1.6.10**
Installation method: **rpm**

Did I have portmaster before? _Well, sort of. I tried installing it before using rpm and then using the script. Encountered the same issue, and then tried to removed it to the best of my ability._ 

Did I try to re-install: _Yes, doesn't help._

Output of
```sh
❯ ls -lah /etc/systemd/system/portmaster.service
lrwxrwxrwx. 1 root root 41 Dec  6 20:58 /etc/systemd/system/portmaster.service -> /opt/safing/portmaster/portmaster.service
````

**The chain of events**
After doing install I immediately restarted the system and found out that it did not autostart. 

I tried executing the 
```sh
sudo systemctl status portmaster
```

and got this error
```
Unit portmaster.serivce could not be found
```

Same thing goes for
```sh
sudo systemctl start portmaster
```

After that I tried starting portmaster UI. Upon the start, I was greeted with the message "The portmaster service is not running" and the button "START CORE SERVICE". Clicking the button didn't help, because it tries to execute this command
```sh
/bin/bash -c echo SUDOPROMPT; systemctl start portmaster.service --no-pager
```
which obviously encounters the same "Unit can't be found" issue.

So then I attempted to execute 
```sh
sudo systemctl enable portmaster
```
and that didn't return any error. 

After that both `status` and `enable` worked just fine. This screenshot shows the output of the `status` after executing `enable`, but before `start`:
![Screenshot from 2024-12-06 22-16-58](https://github.com/user-attachments/assets/0a31b719-c2d2-481e-a171-81a62c68295d)

And this one after executing `start`:

![Screenshot from 2024-12-06 22-18-55](https://github.com/user-attachments/assets/130e1ca7-2388-4478-8786-10b45a1096d8)

Then, I tried rebooting my machine, but no autostart happened. `enable` didn't find the unit. So I decided to try execute `enable` and `start` in one command and see if that works:
```sh
sudo systemctl enable --now portmaster
```
it did work.

Finally, I rebooted once again, and found myself in the same situation. I tried starting the portmaster using 
```sh
sudo /var/opt/safing/portmaster/portmaster-start core
```
it did start fine as well.

Here's the output of the `journalctl -u portmaster` for the latest boot (that was performed using `systemctl enable`). And the shutdown was imitated by the system reboot.
[log.txt](https://github.com/user-attachments/files/18038719/log.txt)
